### PR TITLE
neutrinordp: Allow users or administrators to configure the mstsc experience settings.

### PR DESCRIPTION
**Pull request by TOMATO-ONE**
Hi.

I am connecting to remote Windows 10 Pro with NeutrinoRDP Proxy.
I have found that when I have mouse cursor shadow enabled, the mouse cursor turns into a black square.

This is similar to what was reported in FreeRDP [issue 5337](https://github.com/FreeRDP/FreeRDP/issues/5337#).
The workaround is to disable the option to `Enable pointer shadow` in the mouse properties of the remote Windows.

I have confirmed that the same can be done with the `xrdp-neutrinordp.c` function `lfreerdp_pre_connect()`.
b8ec98a

Best regards.

### CODE REVIEW

**Comment by matt335672** [comment]
Hi @TOMATO-ONE 

Thanks for raising this.

This has been mentioned before - see #716, #1003, #1333.

Currently, I'm not comfortable with embedding a workaround in code without making an effort to figure out what is going on. I suspect that with a bit of extra effort we can probably get this working between us.

When we implemented #1807 we added some extra logging which may help with this. This logging in in v0.9.16 and the devel version so maybe that will help:-
https://github.com/neutrinolabs/xrdp/blob/148000c46bed3ed6ff0ef4e3a5ac5bc50c770fea/neutrinordp/xrdp-neutrinordp.c#L1455-L1457

Are you seeing any messages like this in the xrdp log file?

**Comment by TOMATO-ONE** [comment]
Hi. @matt335672 

The `new_cursors=false` mentioned in #716 did not work around this.

A fix for a similar issue was made in FreeRDP PR [#4552](https://github.com/FreeRDP/FreeRDP/pull/4552/commits/71c949f61a8ed4584546b5b0ba77ab8f59e40b67), and the fix was beyond my understanding, so I gave up.
As you said, it was hasty to try to implement the workaround without knowing the cause.

In the normal build of xrdp v0.9.16, I got the following DEBUG log.
```
[20210531-20:44:28] [WARN ] lfreerdp_pointer_new: error bpp 32 width 41 height 39 index: 0
[20210531-20:44:29] [DEBUG] Request to send a message to the disabled channel drdynvc (3)
[20210531-20:44:30] [WARN ] lfreerdp_pointer_new: error bpp 32 width 41 height 39 index: 1
[20210531-20:44:30] [DEBUG] Request to send a message to the disabled channel drdynvc (3)
<snip>
[20210531-20:45:09] [DEBUG] Request to send a message to the disabled channel drdynvc (3)
[20210531-20:45:10] [WARN ] lfreerdp_pointer_new: error bpp 32 width 41 height 39 index: 2
[20210531-20:45:16] [WARN ] lfreerdp_pointer_new: error bpp 32 width 41 height 39 index: 3
[20210531-20:45:19] [WARN ] lfreerdp_pointer_new: error bpp 32 width 41 height 39 index: 4
[20210531-20:45:25] [WARN ] lfreerdp_pointer_new: error bpp 32 width 41 height 39 index: 5
[20210531-20:45:50] [WARN ] lfreerdp_pointer_new: error bpp 32 width 41 height 39 index: 7
```

In xrdpdebug build of xrdp devel
```
/ # tail -f /var/log/xrdp.log | grep lfreerdp_pointer_new
[20210531-11:14:16] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1399)] lfreerdp_pointer_new:
[20210531-11:14:16] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1400)]   bpp 32
[20210531-11:14:16] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1401)]   width 41 height 39
[20210531-11:14:16] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1404)]   lengthXorMask 6396 lengthAndMask 234
[20210531-11:14:16] [WARN ] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1455)] lfreerdp_pointer_new: error bpp 32 width 41 height 39 index: 0
[20210531-11:14:18] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1399)] lfreerdp_pointer_new:
[20210531-11:14:18] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1400)]   bpp 24
[20210531-11:14:18] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1401)]   width 32 height 32
[20210531-11:14:18] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1404)]   lengthXorMask 3072 lengthAndMask 128
[20210531-11:14:18] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1441)] lfreerdp_pointer_new: bpp 24 Bpp 3
[20210531-11:14:19] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1399)] lfreerdp_pointer_new:
[20210531-11:14:19] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1400)]   bpp 24
[20210531-11:14:19] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1401)]   width 32 height 32
[20210531-11:14:19] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1404)]   lengthXorMask 3072 lengthAndMask 128
[20210531-11:14:19] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1441)] lfreerdp_pointer_new: bpp 24 Bpp 3
[20210531-11:14:21] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1399)] lfreerdp_pointer_new:
[20210531-11:14:21] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1400)]   bpp 24
[20210531-11:14:21] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1401)]   width 32 height 32
[20210531-11:14:21] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1404)]   lengthXorMask 3072 lengthAndMask 128
[20210531-11:14:21] [DEBUG] [lfreerdp_pointer_new(xrdp-neutrinordp.c:1441)] lfreerdp_pointer_new: bpp 24 Bpp 3
```

**Comment by TOMATO-ONE** [comment]
Should we close this PR and continue the discussion in issue #1333?

**Comment by matt335672** [comment]
That's a useful log.

I've had a look at the code, and a pointer size of 32x32 is pretty baked in to the cursor code.

There's a bit to be done to make this work. So far I've identified the following:-
- At the start of the negotiation with the RDP client, we need to send a [Large pointer capability set structure](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/41323437-c753-460e-8108-495a6fdd68a8) to say we support pointers of up to 96x96 pixels.
- The structures we store pointers in need to be made larger (or dynamically allocated) and the pointer size needs to be stored in these structures.
- All the pointer caching routines need to updated to cope with variable size pointers.

It's all doable, and I think we should probably do it, but it's quite a lot of work. There's probably stuff I've left off of the above list too - it's not an area I'm that familiar with.

I'm happy to look into the above (I understand it's pretty hairy) but it'll take a bit of time to get it working and debugged I'm also conscious of the need to give you a better way forward than modifying the code.

Would you be happy to work on your PR above to make it user-controllable? You could for example add an `enable_cursor_shadow=[true|false]` option to the neutrinordp section. The option would default to 'true', but if set to false, you'd disable that performance option as you've done above.

So you'd add something like the following to `xrdp/xrdp.ini.in`:-

```
; By default NeutrinoRDP supports cursor shadows. If this is giving you problems
; (e.g. cursor is a black rectangle) try disabling cursor shadows by uncommenting
; the following line
#enable_cursor_shadow=false
```

Then you'd modify your code above to check the value of the setting and react appropriately.

Let me know what you think.

**Comment by TOMATO-ONE** [comment]
HI. @matt335672 
Thank you for the detailed instruction on the cause.

I will continue to work on this PR.

I actually realized while considering this PR that the NeutrinoRDP Proxy has a fixed performance setting.
Wallpaper, menu animation, etc. are disabled by default.
https://github.com/neutrinolabs/xrdp/blob/148000c46bed3ed6ff0ef4e3a5ac5bc50c770fea/neutrinordp/xrdp-neutrinordp.c#L1670-L1682


I would like to be able to make this configurable by the RDP client user.
I would also like the RDP Proxy administrator to be able to enforce the settings.

In the xrdp.ini, the following parameters can be specified in the nearinordp section.
```
rdpproxy.disable_wallpaper = [ default | true | false | client ]
rdpproxy.disable_fullwindowdrag = [ default | true | false | client ]
rdpproxy.disable_menuanimations = [ default | true | false | client ]
rdpproxy.disable_theming = [ default | true | false | client ]
rdpproxy.enable_font_smoothing = [ default | true | false | client ]
rdpproxy.enable_desktop_composition = [ default | true | false | client ]
rdpproxy.disable_cursorsettings = [ default | true | false | client ]
rdpproxy.disable_cursor_shadow = [ default | true | false  ] *
```
note: There is a value for `disable_cursor_shadow` in `performance_flag`, but I could not find a UI to control it on the RDP Client side.

The possible values for each parameter are as follows
```
default  : use the default value of xrdp neutrinordp proxy 
true : force  to  true 
false : force to false    
client : use the RDP client setting ( mstsc.exe -> experience )
```
Except for the cursor shadow , I have confirmed that I can get the RDP client settings with `rdp5_performanceflags`.
https://github.com/neutrinolabs/xrdp/blob/148000c46bed3ed6ff0ef4e3a5ac5bc50c770fea/common/xrdp_client_info.h#L83

I would like to discuss the following points.
- In Windows 2012R2 and later, the parameters "theming", "desktop_composition", and "font_smoothing" do not reduce the amount of data communication, so I am concerned that the options dedicated to NeutrinoRDP will become bloated. Should I omit them?
- I thought of the context name "rdpproxy" for identification purposes, since the parameter cannot be used in Xorg or Xvnc. Is this appropriate?
- The names of the parameters are not intuitive, but I'm referring to NeutrinoRDP's `settings.h`.
https://github.com/neutrinolabs/NeutrinoRDP/blob/f7832d6ce8f12f4b999a2a4edc6cc72373bc5168/include/freerdp/settings.h#L27-L37
Another idea is to match the description in the mstsc.exe configuration file (*.rdp).
```
disable_wallpaper
allow_font_smoothing
allow_desktop_composition
disable_full_window_drag
disable_menu_anims
disable_themes
disable_cursor_setting
disable_cursor_shadow *
```
note: `disable_cursor_shadow` is not in the *.rdp file
If you have a better parameter name, please suggest it.
- I would suggest changing the name of the PR, but I can't think of a good, concise sentence. Please suggest

 regards.

**Comment by matt335672** [comment]
Hi @TOMATO-ONE 

You've clearly done a lot of thinking about this.

I think I can clear up one mystery above; the cursor shadow setting is possibly taken from the Windows control panel setting rather than being a separate item in the RDP client.

The performance flags are indeed sent over from the client in the TS_EXTENDED_INFO_PACKET from the client. This is documented in [[MS-RDPBCGR] 2.2.1.11.1.1.1](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/05ada9e4-a468-494b-8694-eb806a0ecc89) which is well worth a read, if you haven't already. In particular there's this section:-

> If the connectionType field of the Client Core Data (section 2.2.1.3.2) is set to CONNECTION_TYPE_AUTODETECT (0x07),
> and the client indicates support for network characteristics detection by specifying the
> RNS_UD_CS_SUPPORT_NETCHAR_AUTODETECT (0x0080) flag in the earlyCapabilityFlags field of the Client Core
> Data, then the server SHOULD ignore the contents of the performanceFlags field if the connection type can be
> determined (using the PDUs specified in section 2.2.14) and SHOULD instead determine an appropriate set of
> performance flags to apply to the remote session based on the detected connection type.

We don't support the performance measuring PDUs, so we won't be doing any of the SHOULD items above.

Also, @metalefty has made a very valid point in the past that many xrdp users don't have English as a first language. So I don't think a mix of enable/disable options is a good way to present this, even though it's in a config file and oddly MS have decided it's a good idea.

On my Windows 10 machine (en-GB), the mstsc.exe client 'Experience' tab presents all these options as 'enable', e.g.:-

performance

- Desktop background
- Font smoothing
- Desktop composition
- Show windows contents while dragging
- Menu and Window animation
- Visual Styles

I think names for these options should follow this approach. Modifying your suggestions slightly we get:-

| Option (xrdp ini file)  | Flag setting |
| :---- | :---- |
| perf.wallpaper | PERF_DISABLE_WALLPAPER (inverted) |
| perf.font_smoothing | PERF_ENABLE_FONT_SMOOTHING |
| perf.desktop_composition | PERF_ENABLE_DESKTOP_COMPOSITION |
| perf.full_window_drag | PERF_DISABLE_FULLWINDOWDRAG (inverted) |
| perf.menu_anims | PERF_DISABLE_MENUANIMATIONS (inverted) |
| perf.themes | PERF_DISABLE_THEMING (inverted) |
| perf.cursor_blink | PERF_DISABLE_CURSORSETTINGS (inverted) |
| perf.cursor_shadow | PERF_DISABLE_CURSOR_SHADOW (inverted) |

One of the advantages is that a user can go through the whole set and set them to 'on' or 'off' in the same way that they can use the tick boxes on the client.

As for setting values, this could work:-

1) Settings above are normally commented out.
2) With commented-out settings, if the user sets CONNECTION_TYPE_AUTODETECT, just go with the current settings. That gives us no change in behaviour.
3) With commented-out settings, if auto-detect isn't set, go with the client settings in TS_EXTENDED_INFO_PACKET
4) Any uncommented settings override settings made above in 2 or 3. Use the function `text2bool()` to get the value of a setting if it's defined.

I'm sure as we work through this you'll think of a better name for the PR - I leave that to you.

BTW, thanks for the interesting discussion!

**Comment by TOMATO-ONE** [comment]
Hi, @matt335672 

Thank you for everything.

As a non-native speaker of English, I am always wondering if my idea of a word is appropriate worldwide.
Thank you for the suggestions for appropriate parameter names.
I would like to adopt your suggestion.

I am working on it now.
`perf.*` configuration part is working.

When I was thinking about the handling of the CONNECTION_TYPE_AUTODETECT flag, I realized that.

If we simply accept the mstsc experience setting with or without this flag, the behavior will change when we upgrade from v0.9.16.
Until now, NeutrinoRDP ignored any experience settings made by RDP client users, but after upgrading, NeutrinoRDP will enable them.
For example, the wallpaper may suddenly appear after an upgrade, and users may be confused by the change in appearance and performance, and administrators may be surprised by the increase in data transfer.


So I would like to add one more parameter to `xrdp.ini` neutrinordp section.
```
; Allow the RDP Client user to set the Experience settings.
perf.allow_experiencesettings=[false|true]
```
When this parameter is commented out, mstsc's experience setting will be ignored by neutrinordp as before.

Please let us know what you think.

**Comment by matt335672** [comment]
It's a very good point.

My own thinking however, is not to complicate it with an extra parameter. We can mention in the release notes that behaviour may change. 

My reason is that because NeutrinoRDP users have to build NeutrinoRDP themselves, these users (like yourself) tend to be more knowledgeable than more casual users of xrdp, and are more likely to read release notes before installing updates.

**Comment by matt335672** [comment]
BTW: Nice title.

**Comment by TOMATO-ONE** [comment]
Hi @matt335672 

Thank you for your suggestions!

You're right.
I think NeutrinoRDP Proxy users will read the release notes carefully.
It may not need complex logic.

However, it is much easier to add the logic and then remove the unneeded logic than to add it later.

Therefore, agreeing with your opinion, once we have included the code to determine the `perf.allow_client_experiencesettings parameter`, we have created the code.

If it is still inappropriate during the review process, I would like to remove it.

Regard.

File: neutrinordp/xrdp-neutrinordp.h
Comment (matt335672) [inline neutrinordp/xrdp-neutrinordp.h]: I don't think this include is necessary here.

File: neutrinordp/xrdp-neutrinordp.c
Comment (matt335672) [inline neutrinordp/xrdp-neutrinordp.c:612]: This comment applies to many sections below.

There's a lot of lines here which aren't really necessary.

Firstly, there's a C idiom that you should be aware of to clear bits in a word. You're doing it by setting the bits and then using XOR to change the value. That will definitely work, but it's not the standard way to do it, which is as follows:-

Standard way to set bits in a word:-
```c
mod->value |= bitmask
```

Standard way to clear bits in a word:-
```c
mod->value &= ~bitmask
```

More info [here](https://stackoverflow.com/questions/3920307) [stackoverflow.com]

With that out the way, I can show you a way to implement your change which needs quite a bit less code, and avoids string copies. Unnecessary string copies in C are the source of many problems.

Rather than storing a string for each parameter, all you need are two words in `struct mod` in `xrdp-neutrinordp.h`:-

```
    int perf_settings_override_mask; /* Performance bits overridden in ini file */
    int per_settings_values_mask; /* Values of overridden performance bits */
```

- The override mask has a bit set for every value that the user has overridden, regardless of the value.
- The value mask has the values of the bits which are actually set in the override mask

Then in the code above, your settings become something like this. Note the difference in the code between a disable setting (PERF_DISABLE_WALLPAPER) and an enable setting (PERF_ENABLE_FONT_SMOOTHING)

```c
    /* Initialise performance settings masks */
    mod->perf_settings_override_mask = 0;
    mod->per_settings_values_mask = 0;
    . . .
    else if (g_strcmp(name, "perf.wallpaper") == 0)
    {
        mod->perf_settings_override_mask |= PERF_DISABLE_WALLPAPER;
        if (!g_text2bool(value))
        {
            mod->per_settings_values_mask |= PERF_DISABLE_WALLPAPER;
        }
    }
    else if (g_strcmp(name, "perf.font_smoothing") == 0)
    {
        mod->perf_settings_override_mask |= PERF_ENABLE_FONT_SMOOTHING;
        if (g_text2bool(value))
        {
            mod->per_settings_values_mask |= PERF_ENABLE_FONT_SMOOTHING;
        }
    }
    . . .
```

Then below, your lines 1770 - 1927 can be simply replaced with:-

```c
/* Clear bits for any overridden performance settings */
instance->settings->performance_flags =
    (instance->settings_performance_flags & ~mod->perf_settings_override_mask);

/* Add in overridden performance settings */
instance->settings->performance_flags |= mod->perf_settings_values_mask;
```

Does that make sense?

**Review by matt335672 — COMMENTED** [review]
Hi @TOMATO-ONE - thanks again for putting effort into this.

We can talk about the allowexperiencesettings later. I've mentioned a way below where you should be able to reduce the size of this change, and remove string copying.

File: neutrinordp/xrdp-neutrinordp.h
Comment (TOMATO-ONE) [reply neutrinordp/xrdp-neutrinordp.h]: Certainly.
There were still some things listed during the trial and error process.
This line will be deleted.

File: neutrinordp/xrdp-neutrinordp.c
Comment (TOMATO-ONE) [reply neutrinordp/xrdp-neutrinordp.c:612]: Hi @matt335672 

Thank you for your very detailed advice on my verbose and immature code.

I appreciate your guidance on bit operations. I will relearn it.

Thank you also for the workaround for copying strings.
When the parameter was false and when I commented it out, the variable value became the same, so I used the method of judging by string.

I will fix the code based on the advice you gave.
It will take some time.

Comment (matt335672) [reply neutrinordp/xrdp-neutrinordp.c:612]: Please feel free to ask any questions if you think I haven't been clear enough, and thanks for doing this.

**Comment by TOMATO-ONE** [comment]
HI @matt335672 
I rewrote the code based on your advice.

File: neutrinordp/xrdp-neutrinordp.c
Comment (TOMATO-ONE) [inline neutrinordp/xrdp-neutrinordp.c]: I made the initialization of the variables to be done within the mod_init function.
Please let me know if there is another appropriate place.

Comment (matt335672) [reply neutrinordp/xrdp-neutrinordp.c]: I think you can just remove these.

The structure is allocated with this line:-

```
    mod = (struct mod *)g_malloc(sizeof(struct mod), 1);
```

This make sure the structure is cleared out first. C99 guarantees that an integer type with all bits set to zero will take the value zero, and this assumption is made all over the place in xrdp.

File: neutrinordp/xrdp-neutrinordp.c
Comment (matt335672) [inline neutrinordp/xrdp-neutrinordp.c]: Better expressed as :-
```c
instance->settings->performance_flags &= ~mod->perf_settings_override_mask;
```

I think I suggested what you've got here - my apologies.

File: xrdp/xrdp.ini.in
Comment (matt335672) [inline xrdp/xrdp.ini.in]: Wrap to 80 columns:-

```
; By default NeutrinoRDP supports cursor shadows. If this is giving
; you problems (e.g. cursor is a black rectangle) try disabling cursor
; shadows by uncommenting the following line.
```

**Review by matt335672 — COMMENTED** [review]
Very minor comments below.

**Comment by matt335672** [comment]
Hi @TOMATO-ONE

Thanks for implementing that.

I'd like to talk about the `perf.allow_experience settings` flag you've got too. I'm not entirely sure I understand what you're doing with it, and talking will help me to get there I think.

At the moment, looking at the code, I think the purpose of the flag is to allow the client to specify the experience settings. Is that right?

File: neutrinordp/xrdp-neutrinordp.c
Comment (TOMATO-ONE) [reply neutrinordp/xrdp-neutrinordp.c]: As you pointed out.
This part worked correctly after deleting it.
Delete it.

File: xrdp/xrdp.ini.in
Comment (TOMATO-ONE) [reply xrdp/xrdp.ini.in]: Oh, I'm sorry.
I'll fix it.

**Comment by TOMATO-ONE** [comment]
Hi. @matt335672 
Thanks for the input.

I was lost for a long time while coding.

I thought about the users who don't want this extra feature.
I thought it would be better to not enable this feature unless the user explicitly sets it to true.

That's the only reason I left the `perf.allow_experience_settings`.

If your policy is to focus on eliminating complexity in your code, I think it is fine to remove it.

Please let me know your thoughts again.

File: neutrinordp/xrdp-neutrinordp.c
Comment (TOMATO-ONE) [reply neutrinordp/xrdp-neutrinordp.c]: I'm sorry.
It's my fault for pasting your mere example.
I will fix it.

**Comment by matt335672** [comment]
I think we can keep the setting, but we should modify the comments you've got in the ini file and the code to make the
intent clearer. I'll add a couple of review comments for discussion.

File: xrdp/xrdp.ini.in
Comment (matt335672) [inline xrdp/xrdp.ini.in]: Suggest the following comment instead:-
```
; By default, performance settings requested by the RDP client are ignored
; and chosen by NeutrinoRDP. Uncomment this line to allow the user to
; select performance settings in the RDP client.
```

File: xrdp/xrdp.ini.in
Comment (matt335672) [inline xrdp/xrdp.ini.in]: Suggest this comment:-

```
; Override any experience setting by uncommenting one or more of the
; following lines.
```

File: neutrinordp/xrdp-neutrinordp.c
Comment (matt335672) [inline neutrinordp/xrdp-neutrinordp.c]: Suggest replacing with:-

```c
    /* auto-detect not yet supported - use default performance settings */
```

**Review by matt335672 — COMMENTED** [review]
Couple of suggestions around `allow_client_experiencesettings`

File: neutrinordp/xrdp-neutrinordp.c
Comment (TOMATO-ONE) [reply neutrinordp/xrdp-neutrinordp.c]: Thank you.  I accept your suggestion.

File: xrdp/xrdp.ini.in
Comment (TOMATO-ONE) [reply xrdp/xrdp.ini.in]: I accept your suggestion.

File: xrdp/xrdp.ini.in
Comment (TOMATO-ONE) [reply xrdp/xrdp.ini.in]: We think it is easy to understand.
I accept your suggestion.

**Comment by TOMATO-ONE** [comment]
HI. @matt335672 

I squashed it and pushed it again.
Thank you for suggesting the correct wording!

**Review by matt335672 — APPROVED** [review]

**Comment by matt335672** [comment]
Thanks for working with me on this @TOMATO-ONE 

@metalefty - I'm happy with this PR. Can you take a quick look at it please before I merge?

File: neutrinordp/xrdp-neutrinordp.c
Comment (metalefty) [inline neutrinordp/xrdp-neutrinordp.c]: Using?

File: xrdp/xrdp.ini.in
Comment (metalefty) [inline xrdp/xrdp.ini.in]: This semicolon is unnecessary.

**Review by metalefty — APPROVED** [review]
I found some trivial points but the code LGTM.

**Comment by matt335672** [comment]
@TOMATO-ONE - fix up @metalefty's comment above and I'll merge this.

Thanks both.

**Comment by TOMATO-ONE** [comment]
Thanks. @matt335672 , @metalefty 
We have solved a problem that has been with us for a year.
I'm happy too.

**Pull request merged**
